### DISCUSSION TITLE
source_url fails when used with gitlab private token

### DISCUSSION THREAD

**Discussion by rmhgan** [mesonbuild/meson#12795] [Q&A] [answered]
Hello!

I'm writing a wrap for a subproject that is trying to download a package from gitlab. My wrap file looks like this:

```
[wrap-file]
directory = libtest
source_url = https://PRIVATE-TOKEN:[obfuscated]@gitlab.domain.com/api/v4/projects/818/packages/generic/libtest/1.0/libtest.zip
source_filename = libtest-1.0.tar.gz
[provide]
dependency_names = TRY
```

The problem I'm running into stems from the url provided. Meson appears to parse the first ':' as the divider between host and port, and I'm getting errors saying that I have a non-numeric port number. I need to use the 'PRIVATE-TOKEN:[obfuscated]' to access our gitlab.

Any recommendations? I've tried variants on the url, including specifying the port explictly, hoping then that meson would parse the command correctly, but it still captures the first instance of the colon.

Thanks for your time.

**Comment by eli-schwartz** [comment, marked as answer]
https://mesonbuild.com/Release-notes-for-0-64-0.html#credentials-from-netrc-for-https-urls

Meson supports reading `~/.netrc` which is usually what people want here anyway. Otherwise it simply passes your url string into python urllib and tries to download it. I'm not sure that actually supports parsing this on its own.

**Reply by rmhgan** [reply]
Thank you Eli. I'm checking that out now, I expect that to work.

**Reply by rmhgan** [reply]
Works very well. Appreciate your time and your assistance!!

**Reply by eli-schwartz** [reply]
Fantastic! Thanks for reporting back.
